code/extensions/EditableFormFieldFreshdeskExtension.php

```diff
@@ -0,0 +1,40 @@
+<?php
+
+class EditableFormFieldFreshdeskExtension extends DataExtension
+{
+    private static $db = [
+        'FreshdeskFieldMapping' => 'Text',
+        'FreshdeskFieldCustom' => 'Boolean',
+    ];
+
+    public function updateCMSFields(FieldList $fields)
+    {
+        $fields->addFieldToTab('Root.Main', new TextField('FreshdeskFieldMapping', 'Freshdesk field mapping:'));
+        $fields->addFieldToTab('Root.Main', new CheckboxField('FreshdeskFieldCustom', 'Freshdesk custom field'));
+    }
+
+    /*
+    * Ensure Freshdesk fields exist via API call
+    */
+    public function validate(ValidationResult $validationResult)
+    {
+        if (!$this->owner->FreshdeskFieldMapping) {
+            return $validationResult->valid();
+        }
+
+        $freshdesk = \FreshdeskAPI::create();
+        $result = $freshdesk->APICall('GET', FRESHDESK_API_BASEURL, '/api/v2/ticket_fields');
```

Comment:
Again, having a form extension know details of the API (endpoints, paths, response codes) sounds like wrong level of abstraction.

Reply:
ack